### PR TITLE
Tunable coupler elements and other significant additions

### PR DESCRIPTION
**Pull request by fedroy**
Warning:
Not all the added code has been adjusted to the new structure and no testing has been added for the new code.
During merge I probably reintroduced some removed code and some annoying commented code.
We can try to address these problems before merge or after, not sure what is best practice.

Niklas Glaser (@GlaserN) implemented a significant portion of the changes below.

Changes summary:

- Added feature to calculate an operator expectation value
- Made separate signal generation chain for each control line
- Wrote new pulse shapes
- Added more devices emualting noise, markovian and non (might need update to new code structure)
- Added fidelities specific for 2 qubit gates with coupler
- Wrote robust optimal control in a new optimizer function
- Added Netzero envelopes
- New flux tunable elements (like symmetric and asymmetric transmons)
- Improved hamiltonian reordering process
- Changed all matrix exponentiations to the tensorflow expm function

### CODE REVIEW

**Comment by lazyoracle** [comment]
@fedroy Any way we can bring the test coverage back up to where it was? Codecov suggests that the coverage will go down by 6.5% if we merge which is quite significant. If you think there isn't much to be done on the tests front, then I will approve and merge this.

**Comment by shaimach** [comment]
Can you elaborate what is the added functionality in terms of
* Robust gates
* non-Markovian noise

**Review by lazyoracle — CHANGES_REQUESTED** [review]
> Warning:
Not all the added code has been adjusted to the new structure and no testing has been added for the new code.
During merge I probably reintroduced some removed code and some annoying commented code.
We can try to address these problems before merge or after, not sure what is best practice.

We definitely want to keep PRs self contained. Which means if there is code that got reintroduced which was previously removed in a refactor/cleanup, we most definitely want to tackle it in this PR before merging it into the `dev` or `master`. This is coming from the idea that those branches should be clean and stable. And for merges with the number of commits as here, we definitely want to squash all of it into a single commit, otherwise it will be extremely difficult to keep track of things. I would thus very strongly recommend that we do all the clean-up in this branch which probably at least includes the following:
* Ensure any code that was previously removed (including commented sections) has not crept back in here
* Ensure all sections of the codebase are up to date with the new structure. Not doing this will affect other features that are simultaneously being developed/merged.
* Ensure at least basic testing is added. Firstly, because we do not want to reduce our test coverage. Secondly, we wont know if someone else comes and messes up all the extensive work done by @fedroy @GlaserN without adequate tests for this PR.

I am tagging this PR to `on-hold` while we keep adding more commits to ensure no residual code creeps back and there are adequate tests.

**Comment by fedroy** [comment]
> Can you elaborate what is the added functionality in terms of * Robust gates * non-Markovian noise

@GlaserN wrote a new version of the c1 optimizer that allows for optimization over multiple experiment instances, with variation in model parameters. This should allow for the implementation of robust optimal control. Niklas could comment further, maybe the way he implemented it is slightly different.

I wrote a noise box in the signal generation that introduces Random telegraph signal noise or sum thereof (leading to 1/f type noise). Niklas also has implemented a general filter function object to include the presence of filters along the line.

**Comment by shaimach** [comment]
Re. new features of robustness, control line noise and filter functions
(IIR?):

These are great.

The growing number of features and capabilities means we need to put some
effort into high-level documentation, describing in general terms the
current features and capabilities, and linking down to API and JSON
detailed documentation.

File: c3/generator/devices.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 import os
```
Comment (nwittler) [inline c3/generator/devices.py:1]: All devices need to conform to the signal chain usage, i.e. implement `process()`. Confirm with tests.

File: c3/libraries/estimators.py [hunk 2]
```diff
@@ -150,10 +152,16 @@ def neg_loglkh_gauss_norm_sum(exp_values, sim_values, exp_stds, shots):
...
 
 @estimator_reg_deco
-def g_LL_prime(exp_values, sim_values, exp_stds, shots):
-    std = tf.sqrt(sim_values*(1-sim_values)/shots)
-    return tf.reduce_mean(((exp_values-sim_values)**2 / std**2 - 1) / 2)
-
+def g_LL_prime(exp_values, sim_values, exp_stds, shots, scale = None, pops = None):
+    if scale:
+        std = [tf.sqrt(pops*(1-pops)/shots)+0.001]*tf.abs(scale)
+        # print(f"{pops=}\n{scale=}\n{shots=}\n{std=}")
```
Comment (nwittler) [inline c3/libraries/estimators.py:158]: Revert to the old one. Make this one user-defined.

File: c3/libraries/fidelities.py [hunk 3]
```diff
@@ -60,27 +61,45 @@ def iswap_transfer(U_dict: dict, index, dims, eval, proj=True):
...
+@fid_reg_deco
+def cphase_comp(U_dict: dict, index, dims, eval, proj=True):
+    U = {}
+    U["CZ"] = U_dict["CZ"]
+    infid = unitary_infid(U, "CZ", [0, 1], dims, proj=proj)
+    return infid
+
+
+@fid_reg_deco
+def lindbladian_cphase_comp_sub(U_dict: dict, index, dims, eval, proj=True):
```
Comment (nwittler) [inline c3/libraries/fidelities.py:95]: Move these functions to the user definitions.

File: c3/libraries/fidelities.py [hunk 4]
```diff
@@ -462,47 +478,68 @@ def lindbladian_average_infid_set(U_dict: dict, index, dims, eval, proj=True):
...
         real_cliffords = evaluate_sequences(U_dict, cliffords_decomp)
     elif num_gates == 2:
         real_cliffords = evaluate_sequences(U_dict, cliffords_decomp_xId)
-    projection = "fulluni"
-    if proj:
-        projection = "wzeros"
-        fid_lvls = 2 ** num_gates
-    ideal_cliffords = perfect_cliffords(str(lvls), proj=projection, num_gates=num_gates)
+    ideal_cliffords = perfect_cliffords(lvls=[2] * num_gates, num_gates=num_gates)
+    # gate = list(U_dict.keys())[0]
```
Comment (nwittler) [inline c3/libraries/fidelities.py:491]: Delete commented code.

File: c3/libraries/fidelities.py [hunk 5]
```diff
@@ -462,47 +478,68 @@ def lindbladian_average_infid_set(U_dict: dict, index, dims, eval, proj=True):
...
+        real_cliffords = evaluate_sequences(U_dict, cliffords_decomp_xId)
+    ideal_cliffords = perfect_cliffords(lvls=[2] * num_gates, num_gates=num_gates)
     fids = []
     for C_indx in range(24):
         C_real = real_cliffords[C_indx]
-        C_ideal = tf_super(tf.Variable(ideal_cliffords[C_indx], dtype=tf.complex128))
-        ave_fid = tf_superoper_average_fidelity(C_real, C_ideal, lvls=fid_lvls)
+        C_ideal = tf_super(tf.constant(ideal_cliffords[C_indx], dtype=tf.complex128))
+        ave_fid = tf_superoper_average_fidelity(C_real, C_ideal, lvls=dims)
+        # real_cliffords = evaluate_sequences(U_dict, cliffords_decomp)
```
Comment (nwittler) [inline c3/libraries/fidelities.py:530]: Delete comments.

File: c3/system/chip.py [hunk 6]
```diff
@@ -235,7 +235,7 @@ def get_Lindbladian(self, dims):
 @dev_reg_deco
 class SymmetricTransmon(PhysicalComponent):
```
Comment (fedroy) [inline c3/system/chip.py:236]: Make symmetric transmon a special case of Asymmetric transmon

File: c3/system/model.py [hunk 7]
```diff
@@ -329,12 +352,12 @@ def get_Frame_Rotation(self, t_final: np.float64, freqs: dict, framechanges: dic
             num_oper = tf.Variable(
                 np.matmul(ann_oper.T.conj(), ann_oper), dtype=tf.complex128
             )
-            # TODO test this
-            if self.dressed:
-                num_oper = tf.matmul(
-                    tf.matmul(tf.linalg.adjoint(self.transform), num_oper),
-                    self.transform,
-                )
+            # TODO test dressing of FR
```
Comment (fedroy) [inline c3/system/model.py:355]: leave todo, move to issue if needed but remove commented code below.

File: c3/utils/utils.py [hunk 8]
```diff
@@ -50,8 +50,10 @@ def replace_symlink(path, alias):
         os.remove(alias)
     except FileNotFoundError:
         pass
-    os.symlink(path, alias)
-
+    try:
+        os.symlink(path, alias)
+    except FileExistsError:
+        pass
```
Comment (fedroy) [inline c3/utils/utils.py:56]: fix it in windows?

**Review by fedroy — COMMENTED** [review]
Minor cleanups

**Review by nwittler — CHANGES_REQUESTED** [review]
- Some minor cleanup to be done where indicated. 
- new devices to be made compatible with signal chain
- more tests

After merge:
- Model learning code needs some love

**Comment by lazyoracle** [comment]
The branch is out of date with `dev` (since the SNAIL PR), needs to be rebased on `dev`. @fedroy had been working on the private repo branch, so might be useful to switch development to this branch instead to stay in sync

**Review by nwittler — COMMENTED** [review]
Looks good.

File: c3/generator/devices.py (hunk 1 above)
Comment (nwittler) [reply c3/generator/devices.py:1]: We'll do this in dev.

**Review by nwittler — DISMISSED** [review]
Most significant things fixed. Some larger issues are left for fixing in dev.

**Review by lazyoracle — CHANGES_REQUESTED** [review]
Several tests are failing, at import level, even before running the tests. Would need to fix the tests or the code before its ready to merge.

**Review by lazyoracle — APPROVED** [review]

**Review by nwittler — APPROVED** [review]
Let's keep the precision issue in mind.

**Pull request merged**